src/transform/src/attribute/analysis.rs

```diff
@@ -0,0 +1,441 @@
+    pub struct Derived {
+        /// A record of active analyses and their results, indexed by their type id.
+        analyses: BTreeMap<TypeId, Box<dyn AnalysisBundle>>,
+        /// Analyses ordered where each depends only on strictly prior analyses.
+        order: Vec<TypeId>,
+        /// Map from local identifier to result offset for analysis values.
+        bindings: BTreeMap<LocalId, usize>,
+    }
+
+    impl Derived {
+        /// Return the analysis results derived so far.
+        pub fn results<A: Analysis>(&self) -> Option<&[A::Value]> {
+            let type_id = TypeId::of::<Bundle<A>>();
+            if let Some(bundle) = self.analyses.get(&type_id) {
+                if let Some(bundle) = bundle.as_any().downcast_ref::<Bundle<A>>() {
+                    return Some(&bundle.results[..]);
+                }
+            }
+            None
+        }
+        /// Bindings from local identifiers to result offsets for analysis values.
+        pub fn bindings(&self) -> &BTreeMap<LocalId, usize> {
+            &self.bindings
+        }
+        /// Result offsets for the state of a various number of children of the current expression.
+        ///
+        /// The integers are the zero-offset locations in the `SubtreeSize` analysis,
+        /// which if absent will cause the method to return `None`. The order of the children
+        /// is descending, from last child to first, because of how the information is laid out,
+        /// and the non-reversibility of the look-ups.
+        ///
+        /// It is an error to call this method with more children than expression has
+        pub fn children_of_rev<'a>(
+            &'a self,
+            start: usize,
+            count: usize,
+        ) -> Option<impl Iterator<Item = usize> + 'a> {
+            if let Some(sizes) = self.results::<SubtreeSize>() {
+                let offset = 1;
+                Some((0..count).scan(offset, move |offset, _| {
+                    let result = start - *offset;
+                    *offset += sizes[result];
+                    Some(result)
+                }))
+            } else {
+                None
+            }
+        }
+    }
+
+    /// A builder wrapper to accumulate announced dependencies and construct default state.
+    #[allow(missing_debug_implementations)]
+    #[derive(Default)]
+    pub struct DerivedBuilder {
+        result: Derived,
+    }
+
+    impl DerivedBuilder {
+        /// Announces a dependence on an analysis `A`.
+        ///
+        /// This ensures that `A` will be performed, and before any analysis that
+        /// invokes this method.
+        pub fn require<A: Analysis>(&mut self) {
+            let type_id = TypeId::of::<Bundle<A>>();
+            if !self.result.analyses.contains_key(&type_id) {
+                A::announce_dependencies(self);
+                if self.result.analyses.contains_key(&type_id) {
+                    // panic!("Cyclic dependency detected: {:?}");
```

Comment:
Can we uncomment this panic before merging this?

Reply:
Yes, though I think it's commented because I don't think the logic is correct yet. I'll look into it and see why I thought that, and resolve it one way or another.

Reply:
The logic was .. changed. And I think might now be correct.